File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
struct A {
   int f, t, c;
   bool operator<(A &p){
      return c < p.c;// 오름차순
   }
};
A a[100001];

int p[10001];
int Find(int x){
    if(x == p[x])return x; // 부모 찾음
    else return p[x] = Find(p[x]); // 부모 못찾음
}
void Union(int x, int y) {
   x = p[x];
   y = p[y];
   p[y] = p[x];
}

int main() {
   ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

   int n, m;
   cin >> n >> m;

   for (int i = 1; i <= n; i++)p[i] = i; // make-set
		
	 // 연결
   for (int i = 0; i < m; i++) {
      cin >> a[i].f >> a[i].t >> a[i].c;
   }

   sort(a, a + m); // 정렬

   int ret = 0;
   for (int i = 0; i < m; i++){ // 간선들을...

      int t1 = Find(a[i].f), t2 = Find(a[i].t);

      if (t1 != t2){ // 다른그룹 이면 -> Union
         ret += a[i].c;
         Union(t1, t2);
      }
   }
   
   cout << ret;
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.util.*;

class Main {

    static class Edge implements Comparable<Edge> {
        int from, to, c;

        public Edge(int from, int to, int c) {
            this.from = from;
            this.to = to;
            this.c = c;
        }

        @Override
        public int compareTo(Edge other) {
            return Integer.compare(this.c, other.c);
        }
    }

    static int[] p;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        int m = scanner.nextInt();

        p = new int[n + 1];
        for (int i = 1; i <= n; i++) p[i] = i;

        Edge[] edges = new Edge[m];
        for (int i = 0; i < m; i++) {
            int from = scanner.nextInt();
            int to = scanner.nextInt();
            int c = scanner.nextInt();
            edges[i] = new Edge(from, to, c);
        }

        Arrays.sort(edges);

        int ret = 0;
        for (Edge edge : edges) {
            int t1 = find(edge.from);
            int t2 = find(edge.to);

            if (t1 != t2) {
                ret += edge.c;
                union(t1, t2);
            }
        }

        System.out.println(ret);
    }

    static int find(int x) {
        if (x == p[x]) return x;
        return p[x] = find(p[x]);
    }

    static void union(int x, int y) {
        x = find(x);
        y = find(y);
        p[y] = p[x];
    }
}
